upup/pkg/fi/assettasks/copyfile.go

```diff
@@ -0,0 +1,203 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package assettasks
+
+import (
+	"fmt"
+
+	"net/url"
+
+	"bytes"
+
+	"github.com/golang/glog"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/util/pkg/hashing"
+	"k8s.io/kops/util/pkg/vfs"
+	"strings"
+)
+
+// CopyFile copies an from a source file repository, to a target repository,
+// typically used for highly secure clusters.
+//go:generate fitask -type=CopyFile
+type CopyFile struct {
+	Name       *string
+	SourceFile *string
+	TargetFile *string
+	// @justinsb not sure I like this but the problem is knowing if the sha is a file or a string
+	// so I did a location or a sha.  We have both SHAs in remote file locations, and SHAs in strings
+	SourceShaLocation *string
+	SourceSha         *string
+	TargetSha         *string
+	Lifecycle         *fi.Lifecycle
+}
+
+var _ fi.CompareWithID = &CopyFile{}
+
+func (e *CopyFile) CompareWithID() *string {
+	return e.Name
+}
+
+func (e *CopyFile) Find(c *fi.Context) (*CopyFile, error) {
+	// TODO do a head call on the file??
+	return nil, nil
+
+}
+
+func (e *CopyFile) Run(c *fi.Context) error {
+	return fi.DefaultDeltaRunMethod(e, c)
+}
+
+func (s *CopyFile) CheckChanges(a, e, changes *CopyFile) error {
+	if fi.StringValue(e.Name) == "" {
+		return fi.RequiredField("Name")
+	}
+	if fi.StringValue(e.SourceFile) == "" {
+		return fi.RequiredField("SourceFile")
+	}
+	if fi.StringValue(e.TargetFile) == "" {
+		return fi.RequiredField("TargetFile")
+	}
+	return nil
+}
+
+func (_ *CopyFile) Render(c *fi.Context, a, e, changes *CopyFile) error {
+
+	source := fi.StringValue(e.SourceFile)
+	target := fi.StringValue(e.TargetFile)
+	sourceSha := fi.StringValue(e.SourceSha)
+	sourceSHALocation := fi.StringValue(e.SourceShaLocation)
+
+	glog.Infof("copying bits from %q to %q", source, target)
+
+	if err := transferFile(source, target, sourceSha, sourceSHALocation); err != nil {
+		return fmt.Errorf("unable to transfer %q to %q: %v", source, target, err)
+	}
+
+	return nil
+}
+
+func transferFile(source string, target string, sourceSHA string, sourceSHALocation string) error {
+	data, err := vfs.Context.ReadFile(source)
```

Comment:
Do you want me to write the file to disk?  I think it needs to be in memory in order to compare the SHAs anyways.